File: 198/main.py
from typing import List


class Solution:
    def rob(self, nums: List[int]) -> int:
        """
        approach
        dp
        each point is the max we can get up to that house
        for each house, we can have 2 cases, either we rob it or not
        if we rob it, our max is curr + dp[i - 2]
        if we don't rob it, our max is curr + dp[i - 1]
        return dp[-1]
        """
        n = len(nums)
        dp = [0] * (n + 2)
        for i in range(n):
            dp[i] = max(nums[i] + dp[i - 2], dp[i - 1])
        return dp[-3]

File: 198/main.ts
function rob(nums: number[]): number {
  let maxSums = [0, 0];
  for (let i = 0; i < nums.length; i++) {
    maxSums.push(Math.max(nums[i] + maxSums[i], maxSums[i + 1]));
  }
  return maxSums[maxSums.length - 1];
}

// for each house, we decide if we can rob it or not
// if rob, total = current + nums[i - 2]
// if not rob, total = nums[i - 1]
//
